src/main/java/com/apicatalog/jsonld/flattening/NodeMapBuilder.java

```diff
@@ -98,23 +99,66 @@ public NodeMap build() throws JsonLdError {
 
         // 1.
         if (JsonUtils.isArray(element)) {
+            return handle1();
+        }
 
-            // 1.1.
-            for (JsonValue item : element.asJsonArray()) {
+        // 2.
+        final Map<String, JsonValue> elementObject = new LinkedHashMap<>(element.asJsonObject());
 
-                JsonStructure itemValue;
+        // 3.
+        if (elementObject.containsKey(Keywords.TYPE)) {
+            handle3(elementObject);
+        }
 
-                if (JsonUtils.isObject(item)) {
-                    itemValue = item.asJsonObject();
+        // 4.
+        if (elementObject.containsKey(Keywords.VALUE)) {
+            handle4(elementObject);
+        }
 
-                } else if (JsonUtils.isArray(item)) {
-                    itemValue = item.asJsonArray();
+        // 5.
+        else if (elementObject.containsKey(Keywords.LIST)) {
+            handle5(elementObject);
+        }
```

Comment:
The first thing I've done here is to split the code into multiple methods so it's easier to see what code is taking the most time to execute.